dll/win32/iphlpapi/iphlpapi_main.c

```diff
@@ -775,7 +775,7 @@ DWORD WINAPI GetBestRoute(DWORD dwDestAddr, DWORD dwSourceAddr, PMIB_IPFORWARDRO
         DWORD numShifts, mask;
 
         for (numShifts = 0, mask = table->table[ndx].dwForwardMask;
-         mask && !(mask & 1); mask >>= 1, numShifts++)
+         mask > 0; numShifts += (mask & 1), mask >>= 1)
```

Comment:
This doesn't seem right. You're turning a longest prefix match into a "most bits matched". Any concrete examples of broken values that don't yield the right result as-is?

Reply:
I agree the original code does seem broken too.

Routing table example:

```
Network Destination        Netmask          Gateway       Interface  Metric
          0.0.0.0          0.0.0.0    192.168.178.1   192.168.178.70     25
        127.0.0.0        255.0.0.0         On-link         127.0.0.1    331
        127.0.0.1  255.255.255.255         On-link         127.0.0.1    331
```

The idea here is that for 127.0.0.1 the second route should get chosen, because it is a more specific match.
This should be done by counting the '1's in the netmask starting from the MSB.

The current code seems to count zeros starting at the LSB, which of course would make a "shortest prefix match".

Reply:
> I agree the original code does seem broken too.

+1

> The idea here is that for 127.0.0.1 the last route should get chosen, because it is the most specific match.

So 127.0.0.1 should be matched to `127.0.0.0        255.0.0.0` and not the last?

I think I got it... rethinking the algo...

Reply:
No, 127.0.0.1 should be matched to the last route, because that has a 32 bit prefix match as opposed to an 8 bit one.

Reply:
Hmm... but that's is what it will do now:

numShifts would be equal to 8 for `127.0.0.0 255.0.0.0`
and equal to 32 for `127.0.0.1  255.255.255.255`

32 > 8, so last index would be chosen.

Binary representation of the mask is: 
```cpp
mask = 0xFFFFFF; // 255.255.255.0 | FF FF FF 00
mask = 0xFFFF;   // 255.255.0.0 | FF FF 00 00
mask = 0xFF;     // 255.0.0.0 | FF 00 00 00
```

Reply:
Yes, your version counts "most bits", which with a well-formed netmask is the same thing. I don't know if a netmask like FF000001 is supposed to do anything (or just error out), but it's that kind of value that your algorithm would behave differently for.

Reply:
How about we just use `hostMaskSize += _BitScanForward(&hostMaskSize , table->table[ndx].dwForwardMask)` and then look for the minimum hostMaskSize, and avoid this loop altogether? ;)

Reply:
FF000001 (255.0.0.1) would not be corrent network mask... but I don't know how to use _BitScanForward in this case.

Can you provide some example?

Reply:
Unless I messed up, the _BitScanForward snippet should do the right thing already. It should put 0 in hostMaskSize for FFFFFFFF, 1 for FFFFFFFE, 31 for 80000000, 32 for 00000000. You can try it out with some example values

Reply:
How about this variation?
```cpp
        DWORD numShifts = 0, mask;

        mask = table->table[ndx].dwForwardMask;
#if BYTE_ORDER == LITTLE_ENDIAN
        mask = _byteswap_ulong(mask);
#endif
        while (((mask >> 31) & 1) && numShifts < 32)
        {
            mask <<= 1;
            numShifts++;
        }
        if (numShifts > matchedBits) {
          matchedBits = numShifts;
          matchedNdx = ndx;
        }
```
(anyway I'll test _BitScanForward now...)

Reply:
Looks like `_BitScanForward` doesn't work... `hostMaskSize` is always = 0
```cpp
result = _BitScanForward(&hostMaskSize, 0);
FIXME("/0: %d, %d\r\n", hostMaskSize, result);
result = _BitScanForward(&hostMaskSize, 0xFF);
FIXME("/8: %d, %d\r\n", hostMaskSize, result);
result = _BitScanForward(&hostMaskSize, 0xFFFF);
FIXME("/16: %d, %d\r\n", hostMaskSize, result);
result = _BitScanForward(&hostMaskSize, 0xF0FFFF);
FIXME("/20: %d, %d\r\n", hostMaskSize, result);
result = _BitScanForward(&hostMaskSize, 0xFFFFFF);
FIXME("/24: %d, %d\r\n", hostMaskSize, result);
result = _BitScanForward(&hostMaskSize, 0xFFFFFFFF);
FIXME("/32: %d, %d\r\n", hostMaskSize, result);

And the output:

fixme:(../dll/win32/iphlpapi/iphlpapi_main.c:778) /0: 0, 0
fixme:(../dll/win32/iphlpapi/iphlpapi_main.c:780) /8: 0, 1
fixme:(../dll/win32/iphlpapi/iphlpapi_main.c:782) /16: 0, 1
fixme:(../dll/win32/iphlpapi/iphlpapi_main.c:784) /20: 0, 1
fixme:(../dll/win32/iphlpapi/iphlpapi_main.c:786) /24: 0, 1
fixme:(../dll/win32/iphlpapi/iphlpapi_main.c:788) /32: 0, 1
fixme:(../dll/win32/iphlpapi/iphlpapi_main.c:778) /0: 0, 0
fixme:(../dll/win32/iphlpapi/iphlpapi_main.c:780) /8: 0, 1
fixme:(../dll/win32/iphlpapi/iphlpapi_main.c:782) /16: 0, 1
fixme:(../dll/win32/iphlpapi/iphlpapi_main.c:784) /20: 0, 1
fixme:(../dll/win32/iphlpapi/iphlpapi_main.c:786) /24: 0, 1
fixme:(../dll/win32/iphlpapi/iphlpapi_main.c:788) /32: 0, 1
```

Reply:
Of course it's always 0 in your example, you always have the LSB set. You'll want to test 0xff000000, 0xffff0000, 0xffffff00 instead.

Reply:
Thanks @ThFabba, that did the trick!